modules/network/src/main/java/org/apache/ignite/internal/network/serialization/marshal/ProxyMarshaller.java

```diff
@@ -0,0 +1,122 @@
+        }
+
+        field.setAccessible(true);
+
+        return field;
+    }
+
+    ProxyMarshaller(ValueWriter<Object> valueWriter, ValueReader<Object> valueReader) {
+        this.valueWriter = valueWriter;
+        this.valueReader = valueReader;
+    }
+
+    boolean isProxyClass(Class<?> classToCheck) {
+        return Proxy.isProxyClass(classToCheck);
+    }
+
+    void writeProxy(Object proxy, DataOutputStream output, MarshallingContext context) throws MarshalException, IOException {
+        assert Proxy.isProxyClass(proxy.getClass());
+
+        BuiltInMarshalling.writeClassArray(proxy.getClass().getInterfaces(), output, context);
+
+        valueWriter.write(Proxy.getInvocationHandler(proxy), output, context);
+    }
+
+    Object preInstantiateProxy(DataInputStream input, UnmarshallingContext context) throws UnmarshalException, IOException {
+        Class<?>[] interfaces = BuiltInMarshalling.readClassArray(input, context);
+
+        return Proxy.newProxyInstance(context.classLoader(), interfaces, placeholderInvocationHandler);
+    }
+
+    void fillProxyFrom(DataInputStream input, Object proxyToFill, UnmarshallingContext context) throws UnmarshalException, IOException {
+        InvocationHandler invocationHandler = readInvocationHandler(input, context);
+        replaceInvocationHandler(proxyToFill, invocationHandler);
+    }
+
+    private InvocationHandler readInvocationHandler(DataInputStream input, UnmarshallingContext context)
+            throws IOException, UnmarshalException {
+        Object object = valueReader.read(input, context);
+        if (!(object instanceof InvocationHandler)) {
+            throw new UnmarshalException("Expected an InvocationHandler, but read " + object);
+        }
+        return (InvocationHandler) object;
+    }
+
+    private void replaceInvocationHandler(Object objectToFill, InvocationHandler invocationHandler) throws UnmarshalException {
+        try {
+            proxyHandlerField.set(objectToFill, invocationHandler);
+        } catch (IllegalAccessException e) {
+            throw new UnmarshalException("Cannot set Proxy.h", e);
+        }
+    }
+
+    private static class PlaceholderInvocationHandler implements InvocationHandler {
+        /** {@inheritDoc} */
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            if ("equals".equals(method.getName()) && method.getParameterTypes().length == 1
+                    && method.getParameterTypes()[0] == Object.class) {
+                return proxy == args[0];
+            }
+            if ("hashCode".equals(method.getName()) && noArgs(method)) {
+                return hashCode();
+            }
+            if ("toString".equals(method.getName()) && noArgs(method)) {
+                return "Proxy with placeholder";
+            }
+
+            throw new UnsupportedOperationException("This is a dummy placeholder, but it was got an invocation on " + method);
```

Comment:
Not sure if this error makes sense for a user. Maybe just return null? This handler should not be called anyway

Reply:
If we just return `null`, it may cause another exception in another place that will mask the original problem. On the other hand, if we throw an exception ourselves (as we know that we should not reach that point of execution), this makes the problem clear to a person who encounters it.

Reply:
Ok, then how about just throwing an exception?

Reply:
On line 115, an exception is thrown :)